### PR TITLE
Vectorize function calls

### PR DESCRIPTION
**Pull request by rlouf**
In this PR I introduce a `outlines.vectorize` function that works similarly to `numpy.vectorize`, and allows to turn functions into functions that take arrays as an input. The major differences with `numpy.vectorize` are:

1. The possibility to batch the execution. Launching too many concurrent calls at the same time, doing inference locally with a batch size too large may not be acceptable and thus we need to batch the parallel calls.
2. It will execute coroutines concurrently for better performance.

This function will be used internally for model calls, so we can pass arrays of arbitrary shapes as inputs and get an array of answers. Usual broadcasting rules apply:

```python
import outlines.models as models

complete = models.text_completion.openai("text-davinci-003")

prompts = np.array([["Hi", "Hello"], ["Bye", "Ciao"]])
answer = complete(prompts)
print(answer.shape)
# (2, 2)

answer_samples = complete(prompts, num_samples=10)
print(answer_samples.shape)
# (2, 2, 10)
```

I made the choice here to execute the coroutines in a new event loop before returning. This allows to keep the synchronous user-facing API for now. Of course, this restricts the potential speedups we get by using `asyncio` and we will eventually need to move towards an explicitly async API. But that would be too big of a change for a single PR.

- [x] Vectorize scalar functions and coroutines (OpenAI API, simple tools)
- [x] Vectorize functions with an arbitrary signature (HF models)
- [x] Vectorize coroutines with an arbitrary signature (OpenAI with several samples)
- [x] Execute coroutines in a new event loop before returning
- [x] Replace the OpenAI calls with async ones and use `vectorize`
- [x] Use `vectorize` for HF model calls

### CODE REVIEW

File: outlines/base.py [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+                )
+
+            outputs = _update_arrays_type(outputs, results)
+
+            for output, result in zip(outputs, results):
+                output[index] = result
+
+        return outputs[0] if nout == 1 else outputs
+
+    def vectorize_call(self, broadcast_shape, flat_args, flat_kwargs):
```
Comment (rlouf) [inline outlines/base.py:153]: We don't need to duplicate the code between `vectorize_call` and `vectorize_call_coroutine`

File: outlines/base.py [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+                    np.asarray(x).reshape(broadcast_shape).squeeze()
+                    for x in zip(*outputs)
+                ]
+            )
+        else:
+            outputs = np.asarray(outputs).reshape(broadcast_shape).squeeze()
+
+        return outputs
+
+    def call_with_signature(self, *args, **kwargs):
```
Comment (rlouf) [inline outlines/base.py:99]: We should document this function for future contributors as this is not completely trivial

File: outlines/base.py [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+
+            if n_results == 1:
+                results = (results,)
+
+            if nout != n_results:
+                raise ValueError(
+                    f"wrong number of outputs from the function, expected {nout}, got {n_results}"
+                )
+
+            if outputs is None:
```
Comment (rlouf) [inline outlines/base.py:138]: Perform these operations outside of the loop so we can run coroutine concurrently

**Comment by rlouf** [comment]
Update: Vectorization is currently implemented, although it needs more documentation and tests to make the PR mergeable.

File: outlines/base.py [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+
+            outputs = await asyncio.gather(*tasks)
+
+            return outputs
+
+        loop = asyncio.new_event_loop()
+        try:
+            outputs = loop.run_until_complete(create_and_gather_tasks())
+        finally:
+            loop.close()
```
Comment (brandonwillard) [inline outlines/base.py:245]: Same consideration here.

File: outlines/base.py [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+    """
+
+    def __init__(self, func, signature=None):
+        self.func = func
+        self.signature = signature
+        self.is_coroutine_fn = inspect.iscoroutinefunction(func)
+
+        self.__name__ = func.__name__
+        if hasattr(func, "__doc__"):
+            self.__doc__ = func.__doc__
```
Comment (brandonwillard) [inline outlines/base.py:35]: This makes me wonder if some `functools.wraps`-based approach would work for this as well.

**Review by brandonwillard — APPROVED** [review]
Looks like we could use some tests for OpenAI, but I imagine that's not particularly straightforward.

Aside from that, just some questions/comments; otherwise, looks good to me.

File: outlines/base.py [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+
+        Thunks have no arguments and can thus be called directly.
+
+        """
+        if self.is_coroutine_fn:
+            loop = asyncio.new_event_loop()
+            try:
+                outputs = loop.run_until_complete(self.func())
+            finally:
+                loop.close()
```
[the remark this answers is not in the document]
Comment (rlouf) [reply outlines/base.py:64]: I think the function that returns the new sequence should be async, but I didn't want to break the API just yet.

**Comment by rlouf** [comment]
> Looks like we could use some tests for OpenAI, but I imagine that's not particularly straightforward.

We need to refactor the code to make it easier to test the code that relies on the OpenAI API. That's the thing that annoys me the most currently.

**Comment by rlouf** [comment]
We might want a slightly different behavior for HuggingFace models: in this case we're better off flattening the input arrays so inference can be run in batch. This does not make this PR any less useful, but probably means that we'll need a different mechanism for these models. Will open an issue once this is merged.

File: outlines/base.py (hunk 5 above)
Comment (rlouf) [reply outlines/base.py:35]: Replaced these lines with `functools.update_wrapper(self, func)`.

**Pull request merged**